<center><img src="https://github.com/insaid2018/Term-1/blob/master/Images/INSAID_Full%20Logo.png?raw=true" width="240" height="100" /></center>

## **<center>Capstone Project Group-1012</center>**
## **<center><font color=#154360><u>EMPLOYEE RETENTION ANALYTICS</center>**

---
# **Table of Contents**
---

**1.** [**Introduction**](#Section1)<br>

**2.** [**Problem Statement**](#Section2)<br>
**3.** [**Installing & Importing Libraries**](#Section3)<br>
  - **3.1** [**Installing Libraries**](#Section31)<br>
  - **3.2** [**Upgrading Libraries**](#Section32)<br>
  - **3.3** [**Importing Libraries**](#Section33)<br>

**4.** [**Data Acquisition & Description**](#Section4)<br>
  - **4.1** [**Data Description**](#Section41)
  - **4.2** [**Data Information**](#Section42)
  
**5.** [**Data Pre-Processing**](#Section5)<br>
  - **5.1.** [**Data Pre-Profiling**](#Section51)
  - **5.2.** [**Data Cleaning**](#Section52)     
     - **5.2.1.** [**Drop Duplicate Rows or Features**](#Section521)
     - **5.2.2.** [**Check for Inconsistent Values**](#Section522)
     - **5.2.3.** [**Handling Inconsistent Values**](#Section523)
  - **5.3.** [**Data Merging**](#Section53)
     - **5.3.1.** [**Merging the 3 Dataset**](#Section531)
  - **5.4** [**Saving the Final Dataset to a CSV file**](#Section54)
  
**6.** [**Feature Engineering Steps**](./FeatureEngineering.ipynb)<br>
   - **6.1** [**Handling Missing Values -Univariate**](./FeatureEngineering.ipynb)
   - **6.2** [**Data Splitting**](./FeatureEngineering.ipynb)
   - **6.3** [**Data Encoding**](./FeatureEngineering.ipynb)
   - **6.4** [**Feature Scaling**](./FeatureEngineering.ipynb)
   - **6.5** [**Handling Missing Values -Multivariate**](./FeatureEngineering.ipynb)

**7.** [**Test Data Pre-Processing and Post-Processing**](./FeatureEngineering.ipynb)


**8.** [**Exploratory Data Analysis**](./eda.ipynb)<br>

**9.** [**Summarization**](./eda.ipynb)
  - **9.1** [**Summary**](./eda.ipynb)
  - **9.2** [**Conclusion**](./eda.ipynb)<br>
  - **9.3** [**Actionable Insights**](./eda.ipynb)<br>
    

**10.** [**Model Development and Evaluation**](./Model_Building.ipynb)<br>
  - **10.1** [**Model Development & Evaluation without Tuning**](./Model_Building.ipynb)<br>
  - **10.2** [**Hyperparameter Tuning**](./Model_Building.ipynb)<br>
  - **10.3** [**Model Development & Evaluation after Tuning**](./Model_Building.ipynb)<br>
  - **10.4** [**Cross-Validating the Models**](./Model_Building.ipynb)<br>
  - **10.5** [**Model Comparison**](./Model_Building.ipynb)<br>



---
<a name = Section1></a>
# **1. Introduction**
---

- HR Department of a software company is the client for this project.

- They want to try a new initiative to retain employees.

- The idea is to use data to predict whether an employee is likely to leave.

- Once these employees are identified, HR can be more proactive in reaching out to them before it's too late.

- They only want to deal with the data that is related to permanent employees.

<center><img src="https://projects.insaid.co/capstone2/hr.png" width="700" height="500"></center>

---
<a name = Section2></a>
# **2. Problem Statement**
---

**Current Practice**
- Once an employee leaves, he or she is taken an interview with the name “exit interview” and shares reasons for leaving.
- The HR Department then tries and learns insights from the interview and makes changes accordingly.

**Issues with current Approach :**

- 1. This approach is that it's too haphazard. The quality of insight gained from an interview depends heavily on the skill of the interviewer.
- 2. The second problem is these insights can't be aggregated and interlaced across all employees who have left.
- 3. The third is that it is too late by the time the proposed policy changes take effect.

**Requirement : They want to supplement their exit interviews with a more proactive approach.**

---
<a name = Section3></a>
# **3. Installing & Importing Libraries**
---

<a name = Section31></a>
### **3.1 Installing Libraries**

In [1]:
# Package that is required by pandas profiling
!pip install -q datascience 
# Library to generate basic statistics about data
!pip install -q pandas-profiling      
# Package required to connect to the data source
!pip install mysql-connector-python                                 
# Package required for creating requirements.txt file
!pip install pipreqsnb



In [2]:
!pipreqsnb D:\DS_Projects\capstone2_grp1012\Employee_Retention_Analytics\req

pipreqs  D:\DS_Projects\capstone2_grp1012\Employee_Retention_Analytics\req


<a name = Section32></a>
### **3.2 Upgrading Libraries**

- **After upgrading** the libraries, need to **restart the runtime** to make the libraries in sync. 

- Cell above (3.1) and below (3.2) will not be excuted again after restarting the runtime.

In [3]:
!pip install -q --upgrade pandas-profiling

<a name = Section33></a>
### **3.3 Importing Libraries**

In [1]:
#-------------------------------------------------------------------------------------------------------------------------------
import pandas as pd                                                 # Importing for panel data analysis
from pandas_profiling import ProfileReport                          # Importing Pandas Profiling (To generate Univariate Analysis)
pd.set_option('display.max_columns', None)                          # Unfolding hidden features if the cardinality is high
pd.set_option('display.max_colwidth', None)                         # Unfolding the max feature width for better clearity
pd.set_option('display.max_rows', None)                             # Unfolding hidden data points if the cardinality is high
pd.set_option('mode.chained_assignment', None)                      # Removing restriction over chained assignments operations
pd.set_option('display.float_format', lambda x: '%.5f' % x)         # To suppress scientific notation over exponential values
#-------------------------------------------------------------------------------------------------------------------------------
import numpy as np                                                  # Importing package numpys (For Numerical Python)
#-------------------------------------------------------------------------------------------------------------------------------
import matplotlib.pyplot as plt                                     # Importing pyplot interface using matplotlib
import seaborn as sns                                               # Importing seaborn library for interactive visualization
#--------------------------------------------------------------------------------------------------------------------------------
import warnings                                                     # Importing Package to ignore any warnings
import mysql.connector                                              # Importing mysql connector package to connect to mysql database

---
<a name = Section4></a>
# **4. Data Acquisition and Description**
---

- Three datasets are provided that contain information about past employees and their status (still employed or already left).

 1. **department_data** contains information about each department. The schema of the dataset is as follows:
<br>

|Dataset-1| 
|:--|
| **department_data** | 
<br>

|id| **Feature Name** | **Description** |
|:--|:--|:--|
|01| **dept_id** | Unique Department Code. |
|02| **dept_name** | Name of the Department. |
|03| **dept_head** | Name of the Head of the Department. |

<br>
<br>

2. **employee_details_data** consists of Employee ID, their Age, Gender and Marital Status. The schema of this dataset is as follows:
<br>

|Dataset-2| 
|:--|
| **employee_details_data** | 
<br>

|id| **Feature Name** | **Description** |
|:--|:--|:--|
|01| **employee_id** | Unique ID Number for each employee |
|02| **age** | Age of the employee |
|03| **gender** | Gender of the employee |
|04| **marital_status** | Marital Status of the employee |


<br><br>

3. **employee_data** dataset consists of each employee’s Administrative Information, Workload Information, Mutual Evaluation Information and Status.
<br>

|Dataset-3| 
|:--|
| **employee_data** | 
<br>

|id| **Feature Name** | **Description** |
|:--|:--|:--|
|01| **department** | Department to which the employees belong(ed) to |
|02| **salary** | Salary level with respect to rest of their department |
|03| **tenure** | Number of years at the company |
|04| **recently_promoted** | Was the employee promoted in the last 3 years? |
|05| **employee_id** | Unique ID Number for each employee |
|06| **n_projects** | Number of projects employee has worked on |
|07| **avg_monthly_hrs** | Average number of hours worked per month |
|08| **satisfaction** | Score for employee’s satisfaction with the company (higher is better) |
|09| **last_evaluation** | Score for most recent evaluation of employee (higher is better) |
|10| **filed_complaint** | Has the employee filed a formal complaint in the last 3 years? |
|11| **status** | Current employment status (Employed / Left) |

- **Acquiring data from secondary Datastore**

In [5]:
warnings.simplefilter(action='ignore', category=UserWarning) 


cnx = mysql.connector.connect(user='student', password='student',
                              host='cpanel.insaid.co',
                              database='Capstone2')
try :
    dept_query = "Select * from Capstone2.department_data;"
    dept_df = pd.read_sql(dept_query,cnx)
    emp_det_query = "Select * from Capstone2.employee_details_data;"
    emp_det_df = pd.read_sql(emp_det_query,cnx)
    emp_query = "Select * from Capstone2.employee_data;"
    emp_df = pd.read_sql(emp_query,cnx)
    cnx.close() #close the connection

except Exception as e:
    cnx.close()
    print(str(e))
    
print("The shapes of the three datasets are:")    
print("department_data :",  dept_df.shape)
print("employee_details_data :",  emp_det_df.shape)
print("employee_data :",  emp_df.shape)

The shapes of the three datasets are:
department_data : (11, 3)
employee_details_data : (14245, 4)
employee_data : (14150, 11)


<a name = Section41></a>
### **4.1 Data Description**

- The **description** of various features in the three datasets

### 1. department_data : dept_df

In [6]:
dept_df.head()

dept_id    dept_name              dept_head
0   D00-IT           IT             Henry Adey
1   D00-SS        Sales        Edward J Bayley
2   D00-TP         Temp        Micheal Zachrey
3  D00-ENG  Engineering  Sushant Raghunathan K
4   D00-SP      Support         Amelia Westray

In [7]:
dept_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11 entries, 0 to 10
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   dept_id    11 non-null     object
 1   dept_name  11 non-null     object
 2   dept_head  11 non-null     object
dtypes: object(3)
memory usage: 392.0+ bytes


In [8]:
dept_df.describe()

dept_id dept_name   dept_head
count       11        11          11
unique      11        11          11
top     D00-IT        IT  Henry Adey
freq         1         1           1

**<font color=#196F3D>Observations for department_data**
- <i>dept_df dataset is of **shape (11, 3)**
- <i>No rows have missing values
- <i>No duplicate rows found
- <i>All **3 features** are of **Object Type**


### 2. employee_details_data: emp_det_df

In [9]:
emp_det_df.head()

employee_id  age  gender marital_status
0       113558   43    Male        Married
1       112256   24  Female      Unmarried
2       112586   22  Female      Unmarried
3       108071   36    Male        Married
4       116915   38    Male        Married

In [10]:
emp_det_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14245 entries, 0 to 14244
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   employee_id     14245 non-null  int64 
 1   age             14245 non-null  int64 
 2   gender          14245 non-null  object
 3   marital_status  14245 non-null  object
dtypes: int64(2), object(2)
memory usage: 445.3+ KB


In [11]:
emp_det_df.describe()

employee_id         age
count  14245.00000 14245.00000
mean  112123.05054    32.88993
std     8500.45734     9.97083
min   100101.00000    22.00000
25%   105775.00000    24.00000
50%   111298.00000    29.00000
75%   116658.00000    41.00000
max   148988.00000    57.00000

**<font color=#196F3D>Observations for employee_details_data:**

- <i>dept_df dataset is of **shape (14245 , 4)**
- <i>No rows have missing values
- <i>**2 Features** employee_id and age are of **int Type**
- <i>**2 Features** gender and marital_status are of **Object type**
- <i>Employee as young as 22 years and as old as 57 years have worked with an average age of 32 years.


### 3. employee_data: emp_df

In [12]:
emp_df.head()

avg_monthly_hrs department  filed_complaint  last_evaluation  n_projects  \
0        246.00000       None              NaN          0.86684           6   
1        134.00000       None              NaN          0.55572           2   
2        156.00000     D00-SS          1.00000          0.47408           2   
3        256.00000     D00-SP              NaN          0.96136           6   
4        146.00000     D00-SS              NaN          0.50735           2   

   recently_promoted  salary  satisfaction status  tenure  employee_id  
0                NaN  medium       0.13442   Left 4.00000       124467  
1                NaN     low       0.51104   Left 3.00000       112210  
2                NaN  medium       0.40510   Left 3.00000       126150  
3                NaN     low       0.15297   Left 4.00000       125346  
4                NaN  medium       0.43484   Left 3.00000       113707

In [13]:
emp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14150 entries, 0 to 14149
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   avg_monthly_hrs    14150 non-null  float64
 1   department         13443 non-null  object 
 2   filed_complaint    2046 non-null   float64
 3   last_evaluation    12663 non-null  float64
 4   n_projects         14150 non-null  int64  
 5   recently_promoted  297 non-null    float64
 6   salary             14150 non-null  object 
 7   satisfaction       14000 non-null  float64
 8   status             14150 non-null  object 
 9   tenure             14000 non-null  float64
 10  employee_id        14150 non-null  int64  
dtypes: float64(6), int64(2), object(3)
memory usage: 1.2+ MB


In [14]:
emp_df.describe()

avg_monthly_hrs  filed_complaint  last_evaluation  n_projects  \
count      14150.00000       2046.00000      12663.00000 14150.00000   
mean         199.99435          1.00000          0.71840     3.77830   
std           50.83370          0.00000          0.17311     1.25016   
min           49.00000          1.00000          0.31617     1.00000   
25%          155.00000          1.00000          0.56371     3.00000   
50%          199.00000          1.00000          0.72473     4.00000   
75%          245.00000          1.00000          0.87141     5.00000   
max          310.00000          1.00000          1.00000     7.00000   

       recently_promoted  satisfaction      tenure  employee_id  
count          297.00000   14000.00000 14000.00000  14150.00000  
mean             1.00000       0.62121     3.49936 112080.75025  
std              0.00000       0.25048     1.46258   8748.20286  
min              1.00000       0.04006     2.00000      0.00000  
25%              1.00000       0.45036     3.00000 105772.50000  
50%              1.00000       0.65239     3.00000 111291.50000  
75%              1.00000       0.82493     4.00000 116650.75000  
max              1.00000       1.00000    10.00000 148988.00000

In [15]:
emp_df.isnull().mean()

avg_monthly_hrs     0.00000
department          0.04996
filed_complaint     0.85541
last_evaluation     0.10509
n_projects          0.00000
recently_promoted   0.97901
salary              0.00000
satisfaction        0.01060
status              0.00000
tenure              0.01060
employee_id         0.00000
dtype: float64

**<font color=#196F3D>Observations for employee_data**
    
**Missing Values Percentage**
- <i>dept_df dataset is of **shape (14150 , 11)**
- <i>**tenure**  and **satisfaction** columns have around **1% missing values**
- <i>**department column** has approx. **5% missing values**
- <i>**last_evaluation column** has **10% missing values**
- <i>**filed_complaint column** has approx. **85% missing values**
- <i>**recently_promoted column** has approx. **97% missing values**

- **6 Features** avg_monthly_hrs,filed_complaint,last_evaluation, recently_promoted ,satisfaction, tenure are of continous type
- <i>**2 Features** n_projects ,employee_id are of Discrete Type
- <i>**3 Features** department, status, salary are of Categorical Types


---
<a name = Section5></a>
# **5. Data Pre-Processing**
---


<a name = Section51></a>
### **5.1 Data Pre-Profiling** [Before Merging the 3 Dtasets]

- For **quick analysis** pandas profiling is very handy.

- It generates profile reports from a pandas DataFrame.

- For each feature **statistics** are presented in an interactive HTML report.

In [16]:
profile = ProfileReport(df=dept_df)
profile.to_file(output_file='PreProfReport_Dept.html')
print('Accomplished!')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

Accomplished!


In [17]:
profile = ProfileReport(df=emp_det_df)
profile.to_file(output_file='PreProfReport_EmpDet.html')
print('Accomplished!')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

Accomplished!


In [18]:
profile = ProfileReport(df=emp_df)
profile.to_file(output_file='PreProfReport_Emp.html')
print('Accomplished!')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

Accomplished!


<a name = Section52></a>
### **5.2 Data Cleaning**

- In this section, we will **remove** the **duplicate row** present in the dataset.

<a name = Section521></a>
#### **5.2.1 Check for Duplicate Rows**

In [19]:
dept_df[dept_df.duplicated()]

Empty DataFrame
Columns: [dept_id, dept_name, dept_head]
Index: []

In [20]:
emp_det_df[emp_det_df.duplicated()]

Empty DataFrame
Columns: [employee_id, age, gender, marital_status]
Index: []

In [21]:
dup=emp_df[emp_df.duplicated()]
dup.count()


avg_monthly_hrs      29
department           29
filed_complaint       5
last_evaluation      29
n_projects           29
recently_promoted     0
salary               29
satisfaction         29
status               29
tenure               29
employee_id          29
dtype: int64

In [22]:
emp_df['employee_id'].duplicated().sum()


33

**<font color=#196F3D>Observations for Employee Dataset: emp_df**
- <i> **29 rows** are **Duplicate**
- <i> **5 Employee ids** have **zero value**


<a name = Section522></a>
#### **5.2.2 Drop Duplicate Rows**

In [23]:
## Function Definition to drop duplicate values
# -----------------------------------------------------------
def drop_duplicate(df):
    df.drop_duplicates(inplace=True)
    
    get_index=df[df['employee_id']==0].index
    df.drop(get_index, inplace=True)
    return df

In [24]:
# Calling drop_duplicate(df) function to drop duplicates
# -----------------------------------------------------------
emp_df_unique=drop_duplicate(emp_df)

In [25]:
# Verifying if any Duplicate Values. All values have been removed
emp_df_unique[emp_df_unique.employee_id.duplicated()]

Empty DataFrame
Columns: [avg_monthly_hrs, department, filed_complaint, last_evaluation, n_projects, recently_promoted, salary, satisfaction, status, tenure, employee_id]
Index: []

<a name = Section523></a>
#### **5.2.3 Identify Data Inconsistencies if any in Categorical features of the 3 datasets**

In [26]:
# Department id for Department Dataset
dept_df['dept_id'].value_counts()

D00-IT     1
D00-SS     1
D00-TP     1
D00-ENG    1
D00-SP     1
D00-FN     1
D00-PR     1
D00-AD     1
D00-MN     1
D00-MT     1
D00-PD     1
Name: dept_id, dtype: int64

In [27]:
# Gender for emplojyee_details Dataset
emp_det_df['gender'].value_counts()

Male      9382
Female    4863
Name: gender, dtype: int64

In [28]:
# marital_status for emplojyee_details Dataset
emp_det_df['marital_status'].value_counts()

Unmarried    7283
Married      6962
Name: marital_status, dtype: int64

In [29]:
# department for employee_data Dataset
emp_df_unique['department'].value_counts()

D00-SS     3895
D00-ENG    2573
D00-SP     2108
D00-IT     1152
D00-PD      853
D00-MT      812
D00-FN      722
D00-MN      590
-IT         207
D00-AD      175
D00-PR      173
D00-TP      150
Name: department, dtype: int64

**<font color=#196F3D>Observations**
- <i>**IT has to be replaced with DOO-IT**

In [30]:
# department for status Dataset
emp_df_unique['status'].value_counts()

Employed    10761
Left         3355
Name: status, dtype: int64

<a name = Section524></a>
#### **5.2.4 Handling Inconsistencies**

In [31]:
emp_df_unique['department'].replace('-IT','D00-IT', inplace=True)

- Handled Duplicate Values and Employee ids with '0' value 
- Department column has -IT. It has replaced with D00-IT





<a name = Section53></a>
### **5.3 Data Merging and Cleaning**


<a name = Section531></a>
#### **5.3.1 Merging** the **3 Datasets**

- Merging employee_details dataset and employee_data datasets

In [32]:
# Merging employee_details dataset and employee_data datasets
new_df=emp_det_df.merge(emp_df_unique,on='employee_id' ,how='right')

In [33]:
new_df.shape

(14116, 14)

- Merging new_df (from above) and department_data 

In [34]:
# Merging new_df(employee_details and employee_data) dataset and department_data datasets
merged_df=new_df.merge(dept_df,left_on='department', right_on='dept_id', how='left')

In [35]:
merged_df.shape

(14116, 17)

In [36]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14116 entries, 0 to 14115
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   employee_id        14116 non-null  int64  
 1   age                14116 non-null  int64  
 2   gender             14116 non-null  object 
 3   marital_status     14116 non-null  object 
 4   avg_monthly_hrs    14116 non-null  float64
 5   department         13410 non-null  object 
 6   filed_complaint    2041 non-null   float64
 7   last_evaluation    12629 non-null  float64
 8   n_projects         14116 non-null  int64  
 9   recently_promoted  297 non-null    float64
 10  salary             14116 non-null  object 
 11  satisfaction       13966 non-null  float64
 12  status             14116 non-null  object 
 13  tenure             13966 non-null  float64
 14  dept_id            13410 non-null  object 
 15  dept_name          13410 non-null  object 
 16  dept_head          134

In [37]:
merged_df.isnull().sum()

employee_id              0
age                      0
gender                   0
marital_status           0
avg_monthly_hrs          0
department             706
filed_complaint      12075
last_evaluation       1487
n_projects               0
recently_promoted    13819
salary                   0
satisfaction           150
status                   0
tenure                 150
dept_id                706
dept_name              706
dept_head              706
dtype: int64

**<font color=#196F3D>Observations**
- **merged_df** dataset, holds features for all the 3 Datasets
- It is of **Shape (14116,17)**    
- department_id, department_name and dept_head has equal number of missing values.
- The 4 features of the above dataset department, department_id, department_name and dept_head are of categorical type hholds the same context for our Analysis. *3 of these features* will be **removed**
- **Missing Values in Features**
- Satisfaction       |   150 missing values
- Tenure             |   150 missing values
- department         |   706 missing values
- last_evaluation    |  1487 missing values
- filed_complaint    |  12075 missing values
- recently_promoted  |  13819 missing values

    

- Dropping department id, department name and department head

In [38]:
merged_df.drop(columns=['dept_id','dept_name','dept_head'] , axis=1, inplace=True)

#### Verifying Number of features in the Final Dataset

In [39]:
merged_df.shape

(14116, 14)

<a name = Section54></a>
### **5.4 Saving the Final Dataset to a CSV file**


In [40]:
merged_df.to_csv('emp_merged.csv', index=False)